scipy/linalg/basic.py

```diff
@@ -300,6 +301,81 @@ def solveh_banded(ab, b, overwrite_ab=False, overwrite_b=False, lower=False,
     return x
 
 
+def solve_toeplitz(c_or_cr, b, check_finite=True):
+    """Solve a Toeplitz system using Levinson Recursion
+
+    The Toeplitz matrix has constant diagonals, with c as its first column
+    and r as its first row.  If r is not given, ``r == conjugate(c)`` is
+    assumed.
+
+    Parameters
+    ----------
+    c_or_cr : array_like or tuple of (array_like, array_like)
+        The vector ``c``, or a tuple of arrays (``c``, ``r``). Whatever the
+        actual shape of ``c``, it will be converted to a 1-D array. If not
+        supplied, ``r = conjugate(c)`` is assumed; in this case, if c[0] is
+        real, the Toeplitz matrix is Hermitian. r[0] is ignored; the first row
+        of the Toeplitz matrix is ``[c[0], r[1:]]``.  Whatever the actual shape
+        of ``r``, it will be converted to a 1-D array.
+    b : (M,) or (M, K) array_like
+        Right-hand side in ``T x = b``.
+    check_finite : boolean, optional
+        Whether to check that the input matrices contain only finite numbers.
+        Disabling may give a performance gain, but may result in problems
+        (result entirely NaNs) if the inputs do contain infinities or NaNs.
+
+    Returns
+    -------
+    x : (M,) or (M, K) ndarray
+        The solution to the system ``T x = b``.  Shape of return matches shape
+        of `b`.
+
+    Notes
+    -----
+    The solution is computed using Levinson-Durbin recursion, which is faster
+    than generic least-squares methods, but can be less numerically stable.
+    """
+    # If numerical stability of this algorithim is a problem, a future
+    # developer might consider implementing other O(N^2) Toeplitz solvers,
+    # such as GKO (http://www.jstor.org/stable/2153371) or Bareiss.
+    if isinstance(c_or_cr, tuple):
+        c, r = c_or_cr
+        c = _asarray_validated(c, check_finite=check_finite).ravel()
+        r = _asarray_validated(r, check_finite=check_finite).ravel()
+    else:
+        c = _asarray_validated(c_or_cr, check_finite=check_finite).ravel()
+        r = c.conjugate()
+
+    # Form a 1D array of values to be used in the matrix, containing a reversed
+    # copy of r[1:], followed by c.
+    vals = np.concatenate((r[-1:0:-1], c))
+    if b is None:
+        raise ValueError('illegal value, `b` is a required argument')
+    if vals.shape[0] != (2*b.shape[0] - 1):
+        raise ValueError('incompatible dimensions')
+
+    b = _asarray_validated(b)
+    if np.iscomplexobj(vals) or np.iscomplexobj(b):
+        vals = np.asarray(vals, dtype=np.complex128, order='c')
+        b = b.astype(np.complex128)
+
+    else:
+        vals = np.asarray(vals, dtype=np.double, order='c')
+        b = b.astype(np.double)
```

Comment:
It should.
I'd just do b=asarray(b, dtype=dtype), but I already promised not to bikeshed in this PR.

One thing I do not know is what is the 'official' recommendation on complex types, is it np.complex_ or NP.complex128 etc. Probably what you did is OK